flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java

```diff
@@ -192,6 +192,14 @@ protected void handleStateChange(ConnectionState newState) {
 				break;
 			case RECONNECTED:
 				LOG.info("Connection to ZooKeeper was reconnected. Leader retrieval can be restarted.");
+				try {
```

Comment:
The solution lgtm.

Reply:
I noticed that this class uses a lock when creating the node cache, not sure if it is important or not.

Reply:
also, I am not sure if it is important to add ```client.getConnectionStateListenable().addListener(connectionStateListener);
```

Reply:
> I noticed that this class uses a lock when creating the node cache, not sure if it is important or not.

I'll double check. It's my understanding that all of the curator watches are done single threadedly, so I didn't think I would need locks.

Reply:
> also, I am not sure if it is important to add
> 
> ```
> client.getConnectionStateListenable().addListener(connectionStateListener);
> ```

No, we still have our listener going to for the connection. We only need to invalidate our cache, the client is still fine.

Reply:
I think stop can be called from another thread. So it's probably best to use the lock. I'll add it.